### ISSUE TITLE
Android UI & Build improvements

### ISSUE THREAD

**Issue by skmp** [reicast/reicast-emulator#1093]
Opening this to coordinate work (see #1091, `skmp/android-studio`, `tolis764/android-refactor`)

Sub-items I can think of right now. I've filled in the stuff I've worked on so far with [x]

Purely Android Studio specific
---
- [x] Initial Android Studio migration
- [ ] Full migration to android studio. Not sure the current project is set up nicely, has all build targets, etc. MIPS is dead nowdays?
- [ ] Provide debug/release configurations for ndk
- [x] Migrate travis CI to gradle, remove ant-based building. Update instructions on readme and such

Code modernization
---
- [x] Move to new libraries, new best practices, etc
- [x] Drop support for <4.2 (aka make API 17 minimum)
- [x] Make sure things work with Android 7
- [ ] Investigate crash reporters for aggregated error tracking?

Android UI Improvements
- [x] Fix glitches on the UI (the ones I know are file browser related)
- [ ] Investigate reports about non-working joystick input. Probably related to android 7 issues?
- [ ] Work on a nicer UI, grab icons from discs, etc

Core Improvements
- [ ] Prevent uncontrollable crashing on the native side when things go bad (tm)
- [x] Have somewhat more graceful teardown. I think we always crash on exit right now
- [x] Add some basic support for "pausing" the emulator. We'll probably still get killed in the background.

@LoungeKatt We had discussions w/ @tolis764 to drop android <4.2 to simplify the code base, as well as general UI/UX improvements. That also touches many of the core parts, especially around android -> jni -> core path, and application lifetime management (which is totally FUBAR'd from the native site atm).  Making API 17 minimum sounds good to you as well? 

I'm not fully aware of what needs to happen to move to newer android versions so it'd be great if you can both chip in with your expertise :)

**Comment by LoungeKatt** [comment]
Quite a few of these are covered or in progress by https://github.com/reicast/reicast-emulator/pull/1091

- [ ] Migrate travis CI to gradle, remove ant-based building. Update instructions on readme and such
- [ ] Move to new libraries, new best practices, etc
- [ ] Drop support for <4.2 (aka make API 17 minimum)
- [ ] Make sure things work with Android 7
- [ ] Fix glitches on the UI (the ones I know are file browser related)
- [ ] Investigate reports about non-working joystick input. Probably related to android 7 issues?
(https://github.com/reicast/reicast-emulator/commit/912255e922339685ecd14e7483b80c86a6bc2385)

**Comment by LoungeKatt** [comment]
I currently have search integrated into the error logs, but would love moving logging to an "official" server. I had mentioned that long ago but there was no movement on it.

My version of the UI is a WIP, but the basic shell of converting it to the support library and modern design is there. The missing part is building a local database to integrate disk search. If you have a way to get disk images from the file, that is great. If not, I already started converting the old gamesdb code into something a bit more reliable as an option and a bit less "when it works"

**Comment by skmp** [comment]
Getting the game images will need a couple of days of work at least, see if I have time over the weekend to look at it

**Comment by LoungeKatt** [comment]
Sounds good. I think the core UI should probably wait on that to see what it will need before committing to a specific layout. That could drastically change how and when things are loaded.

**Comment by LoungeKatt** [comment]
![Current Progress](https://s14.postimg.org/k6hx4mdwx/Screen_Shot_2018-03-28_at_10.35.15_AM.png)

Here is how it looks so far. I left the popup visible, but it can be dismissed.
That is running on the x86_64 (labeled, but seemingly x86 only) Phoenix OS

**Comment by skmp** [comment]
Another thing to look at. Apart from libdc, those look like system libs (ffmpeg/libav)

**Comment by LoungeKatt** [comment]
ffmpeg had know relocation issues, which I think have been fixed. Actually, they are the ones that sparked the whole text relocations as a widespread concern. Those are probably from Phoenix OS, but it was more showing that it was no longer a crash and now just a popup that could be dismissed. After the conversion, it isn't completely preventing the app from running anymore.

**Comment by tolis764** [comment]
On my part, this is the list of changes which I have implemented:

- New support and design libraries.
- Set the min sdk version to 17.
- A major redesign to navigation drawer.
- Create vector images.
- File browser screen redesign.
- Some refactors in Main Activity in order become more readable.
- New style(theme).
- Create color class where the project's colors are placed.
- Move some code/declarations from Manifest to gradle file.

**Comment by LoungeKatt** [comment]
@tolis764 Did you have further plans for the navigation drawer / file browser or was it just converting the existing code to the current support library version?

When you refactor MainActivity, try to make sure you don't end up completely removing notes unless they are redundant. In usage info, tagging the variable by type (`@param string`) can just be replaced by tagging it with the variable name (`@param url`).

Are you creating a java color class or just moving them to an individual resource?

**Comment by LoungeKatt** [comment]
I think the current PR is lengthy enough. The only two things not done are more improvements that aren't necessarily vital to functionality or performance.

- [ ] Disk info / database
- [ ] Update cloud save

**Comment by LoungeKatt** [comment]
We had a really long list of issues, some smaller issues that were duplicates, and an unused project. I broke this down into the individual parts, merged in the other issues, and added it to the project to actually track progress.

https://github.com/reicast/reicast-emulator/issues/1129
https://github.com/reicast/reicast-emulator/issues/1128
https://github.com/reicast/reicast-emulator/issues/1127
https://github.com/reicast/reicast-emulator/issues/1126

It seemed silly to open another issue just for submission

- [ ] Rebase against master
- [ ] Pull request
- [ ] Review
- [ ] Merge and close

This is more or less the understood final step

**Comment by LoungeKatt** [comment]
Joysticks still need work and axis config support
Icons from disk is still only theoretical
Build targets and debug configuration are WIP
All the core stuff is still pending

**Comment by AbandonedCart** [comment]
Closed in favor of https://github.com/reicast/reicast-emulator/issues/1297
